packages/frontend/core/src/components/attachment-viewer/utils.ts

```diff
@@ -0,0 +1,109 @@
+import type { RenderOut } from '@affine/core/modules/pdf/workers/types';
+import type { AttachmentBlockModel } from '@blocksuite/affine/blocks';
+import { filesize } from 'filesize';
+
+import { downloadBlob } from '../../utils/resource';
+
+export async function getAttachmentBlob(model: AttachmentBlockModel) {
+  const sourceId = model.sourceId;
+  if (!sourceId) {
+    return null;
+  }
+
+  const doc = model.doc;
+  let blob = await doc.blobSync.get(sourceId);
+
+  if (blob) {
+    blob = new Blob([blob], { type: model.type });
+  }
+
+  return blob;
+}
+
+export async function download(model: AttachmentBlockModel) {
+  const blob = await getAttachmentBlob(model);
+  if (!blob) return;
+
+  await downloadBlob(blob, model.name);
+}
+
+export function renderItem(
+  scroller: HTMLElement | null,
+  className: string,
+  data: RenderOut
+) {
+  if (!scroller) return;
+
+  const item = scroller.querySelector(
+    `[data-index="${data.index}"] > div.${className}`
+  );
+  if (!item) return;
+  if (item.firstElementChild) return;
+
+  const { width, height, buffer } = data;
+
+  const canvas = document.createElement('canvas');
+  const ctx = canvas.getContext('2d');
+  if (!ctx) return;
+
+  const imageData = new ImageData(buffer, width, height);
+
+  canvas.width = width;
+  canvas.height = height;
+  canvas.style.width = '100%';
+  canvas.style.height = '100%';
+
+  ctx.putImageData(imageData, 0, 0);
+
+  item.append(canvas);
+}
+
+export function buildAttachmentProps(model: AttachmentBlockModel) {
+  const isPDF = model.type.endsWith('pdf');
+  const pieces = model.name.split('.');
+  const ext = pieces.pop() || '';
+  const name = pieces.join('.');
+  const size = filesize(model.size);
+  return { model, name, ext, size, isPDF };
+}
+
+/**
+ * Generates a set of sequences.
+ *
+ * 1. when `start` is `0`, returns `[0, .., 5]`
+ * 2. when `end` is `total - 1`, returns `[total - 1, .., total - 5]`
+ * 2. when `start > 0` and `end < total - 1`, returns `[18, 17, 19, 16, 20, 15, 21]`
```

Comment:
what's this for

Reply:
the main scene, after scrolling, generates a sequence to expand the page in the center of the view

Reply:
in the third condition, why the seq nums are not continuous

Reply:
<img width="634" alt="Screenshot 2024-11-11 at 11 25 01" src="https://github.com/user-attachments/assets/c2a68669-f21a-4e3b-a4a7-b9640a9aaa78">